compiler/src/main/kotlin/io/spine/protodata/backend/AstComponents.kt

```diff
@@ -0,0 +1,198 @@
+            .build()
+        type = valueField.type()
+    } else {
+        type = desc.type()
+        when {
+            desc.isRepeated -> list = Empty.getDefaultInstance()
+            desc.realContainingOneof != null -> oneofName = desc.realContainingOneof.name()
+            else -> single = Empty.getDefaultInstance()
+        }
+    }
+    return this
+}
+
+/**
+ * Converts this enum value descriptor into an [EnumConstant] with options.
+ *
+ * @see buildConstant
+ */
+internal fun Descriptors.EnumValueDescriptor.buildConstantWithOptions(
+    declaringType: TypeName,
+    documentation: Documentation
+): EnumConstant {
+    val constant = buildConstant(declaringType, documentation)
+    return constant.toBuilder()
+        .addAllOption(listOptions(options))
+        .build()
+}
+
+/**
+ * Converts this enum value descriptor into an [EnumConstant].
+ *
+ * The resulting [EnumConstant] will not reflect the options on the enum constant.
+ *
+ * @see buildConstantWithOptions
+ */
+internal fun Descriptors.EnumValueDescriptor.buildConstant(
+    declaringType: TypeName,
+    documentation: Documentation
+): EnumConstant {
+    return EnumConstant.newBuilder()
+        .setName(constantName { value = name })
+        .setDeclaredIn(declaringType)
+        .setNumber(number)
+        .setOrderOfDeclaration(index)
+        .setDoc(documentation.forEnumConstant(this))
+        .build()
+}
+
+/**
+ * Converts this method descriptor into an [Rpc] with options.
+ *
+ * @see buildRpc
+ */
+internal fun Descriptors.MethodDescriptor.buildRpcWithOptions(
+    declaringService: ServiceName,
+    documentation: Documentation
+) : Rpc {
+    return buildRpc(declaringService, documentation)
+        .toBuilder()
+        .addAllOption(listOptions(options))
+        .build()
+}
+
+/**
+ * Converts this method descriptor into an [Rpc].
+ *
+ * The resulting [Rpc] will not reflect the method options.
+ *
+ * @see buildRpcWithOptions
+ */
+internal fun Descriptors.MethodDescriptor.buildRpc(
+    declaringService: ServiceName,
+    documentation: Documentation
+) : Rpc {
+    val name = name()
+    val cardinality = cardinality()
+    return Rpc.newBuilder()
+        .setName(name)
+        .setCardinality(cardinality)
+        .setRequestType(inputType.name())
+        .setResponseType(outputType.name())
+        .setDoc(documentation.forRpc(this))
+        .setService(declaringService)
+        .build()
+}
+
+/**
+ * Extracts metadata from this file descriptor, including file options.
+ *
+ * @see toFile
+ */
+internal fun Descriptors.FileDescriptor.toFileWithOptions() =
+    toFile()
+        .toBuilder()
+        .addAllOption(listOptions(options))
+        .build()
+
+/**
+ * Extracts metadata from this file descriptor, excluding file options.
+ *
+ * @see toFileWithOptions
+ */
+internal fun Descriptors.FileDescriptor.toFile() = file {
+    path = path()
+    packageName = `package`
+    syntax = this@toFile.syntax.toSyntaxVersion()
+}
+
```

Comment:
I am not sure whether we can remove this empty line (and not break EOF), but if we can, let's.